Audio extension and RVC/Coqui UI updates/bugfixes

Tony-sama opened the pull request:
## New features:

### Audio extension

#### 1.  Play background music depending on character expression
- fallback to "neutral" expression bgm if classify not load or in group chat
- fallback to a random BGM of asset folder if no BGM in character folder
- Assets BGM are stored in public/assets/bgm, user can add file manually or using assets from [ST-Content](https://github.com/SillyTavern/SillyTavern-Content)
- Character BGM must be inside current sprite folder (example: characters/Seraphina/bgm) and labelled by expression:

![image](https://github.com/SillyTavern/SillyTavern/assets/48798118/51631daa-942b-4170-90d7-2b8644665328)

- BGM is choosed at random among "expression_id.pm3" files

#### 2. Play ambient sounds depending on current background
- Located in public/assets/ambient
- Must have the name of the corresponding background with space replace by "-", example:

![image](https://github.com/SillyTavern/SillyTavern/assets/48798118/cf86c2d8-0eed-4285-b741-fb43813343ed)
- Can be added manually or using the assets extension and picking assets from [ST-Content](https://github.com/SillyTavern/SillyTavern-Content)
**more ambient by Deffcolony coming soon!**

#### 3. UI
- Enable/disable all audio (pause)
- Control BGM/Ambient volume, mute on/off
- Control BGM update cooldown
![image](https://github.com/SillyTavern/SillyTavern/assets/48798118/d30e8b98-4714-498f-9a95-b5ab24533eb1)
 

### RVC
- Expression based RVC model 
 - Require classify module to be active
 - Require a RVC model with the name of the expression in the model folder, default to the first found model otherwise
 Example with models trained by Hyeonse:
![image](https://github.com/SillyTavern/SillyTavern/assets/48798118/15e08489-5da4-4b3a-b267-f24ec611a4f3)

## Updates:
 - RVC UI, redesign for easier usage, added tips description to each options
 - RVC/Coqui added user as potential speaker
 
![image](https://github.com/SillyTavern/SillyTavern/assets/48798118/8463c1d6-100d-44b7-af16-76239602bea1)

## Bugfixes:
- Coqui model state detection when no tts folder in appdata (very first install)
- Coqui corrected model detection for model with "--" in their name

### Assets extension (prototype)
- Fetch official assets from Sillytavern-content repository [ST-Content](https://github.com/SillyTavern/SillyTavern-Content)
- List assets available (grouped by type) and allow to download them
- Only show Asset repository URL and button to connect at load
![image](https://github.com/SillyTavern/SillyTavern/assets/48798118/0ae89b66-e64b-41ad-9d45-2a14792c8ffc)
Only fetch the json if user confirm the connection
![image](https://github.com/SillyTavern/SillyTavern/assets/48798118/1a240019-bd86-44a1-a6b1-21af6ff11db4)
Example of assets list retrieved, clicking on checkbox initiate download to the right place
![image](https://github.com/SillyTavern/SillyTavern/assets/48798118/cb20a0a1-1805-47cf-be01-7197c11fecbb)

Tony-sama committed:
Prototype of RVC+classification pipeline

Tony-sama committed:
Prototype of RVC+classification pipeline

Tony-sama committed:
Merge branch 'staging' of https://github.com/Tony-sama/SillyTavern into staging

Tony-sama committed:
Merge branch 'SillyTavern:staging' into staging

Tony-sama committed:
Merge branch 'SillyTavern:staging' into staging

Tony-sama committed:
Prototype of dynamic background music using expression extension

Tony-sama committed:
Prototype of dynamic background music using expression extension

Tony-sama committed:
Redesigned RVC UI for easier use and intelligibility

Tony-sama committed:
Redesigned RVC UI for easier use and intelligibility

Tony-sama committed:
Merge branch 'SillyTavern:staging' into staging

Tony-sama committed:
Audio module proper extension code. Dedicated Audio extension scripts. Simple UI to adjust bgm/ambient. Load bgm per character from character folder and switch audio dynamically w.r.t. expression. Ambient sound updated when background change.

Tony-sama committed:
Audio module proper extension code. Dedicated Audio extension scripts. Simple UI to adjust bgm/ambient. Load bgm per character from character folder and switch audio dynamically w.r.t. expression. Ambient sound updated when background change.

Tony-sama committed:
Merge branch 'staging' of https://github.com/Tony-sama/SillyTavern into staging

Tony-sama committed:
Merge branch 'SillyTavern:staging' into staging

Tony-sama committed:
Add initialization of audio module key in extension settings.

Tony-sama committed:
Add initialization of audio module key in extension settings.

Tony-sama committed:
Merge branch 'staging' of https://github.com/Tony-sama/SillyTavern into staging

Tony-sama committed:
- RVC
        - Added user to character list for mapping
    - Audio
        - Corrected initialization of extension settings
        - reduce debug logs
    - Coqui:
        - Added user to character list for mapping

Tony-sama committed:
- RVC
        - Added user to character list for mapping
    - Audio
        - Corrected initialization of extension settings
        - reduce debug logs
    - Coqui:
        - Added user to character list for mapping

Tony-sama committed:
- Audio
        - refactored dynamic bgm to handle solo and group chat
        - added bgm cooldown variable, user can custom in the ui

Tony-sama committed:
- Audio
        - refactored dynamic bgm to handle solo and group chat
        - added bgm cooldown variable, user can custom in the ui

Tony-sama committed:
Merge branch 'SillyTavern:staging' into staging

Tony-sama committed:
Moved path of defaulf bgm/ambient sounds

Tony-sama committed:
Moved path of defaulf bgm/ambient sounds

Tony-sama committed:
Merge branch 'SillyTavern:staging' into staging

Tony-sama committed:
cleaning up audio module

Tony-sama committed:
cleaning up audio module

Tony-sama committed:
updated default ST bgm

Tony-sama committed:
updated default ST bgm

Tony-sama committed:
Updated default audio extension parameters

Tony-sama committed:
Updated default audio extension parameters

Tony-sama committed:
Add more ambient

Tony-sama committed:
Add more ambient

Tony-sama committed:
Cleanup audio module debug messages

Tony-sama committed:
Cleanup audio module debug messages

Tony-sama committed:
Merge branch 'SillyTavern:staging' into staging

IkariDevGIT commented:
#501 related

Tony-sama committed:
Start rebase without audio assets

Tony-sama committed:
Merge branch 'staging' of https://github.com/Tony-sama/SillyTavern into staging

Tony-sama committed:
continue rebasing

Tony-sama committed:
Merge branch 'SillyTavern:staging' into staging

Tony-sama commented:
Rebased without the audio assets.

Tony-sama committed:
Adapted audio extension to empty bgm/ambient folder

Tony-sama committed:
Merge branch 'SillyTavern:staging' into staging

Tony-sama committed:
cleanup debug logs

Tony-sama committed:
Merge branch 'staging' of https://github.com/Tony-sama/SillyTavern into staging

Tony-sama committed:
Reorganised to handle different kind of assets, starting with audio bgm/ambient.

Tony-sama committed:
prototype of assets list/download extension

Tony-sama committed:
Handle background image with space in name.

IkariDevGIT commented:
> Rebased without the audio assets.

why?

Tony-sama commented:
For space concern on the ST repository (see discord discussion on #code-contribution). The audio will be stored on https://github.com/SillyTavern/SillyTavern-Content. The idea is that it will be the place for officially supported assets (like the audio bgm/ambient for now). User can download the assets they want using the dedicated menu :)

IkariDevGIT commented:
@Tony-sama When will this be merged into staging?

Cohee1207 commented on public/scripts/extensions/assets/index.js:
```diff
@@ -0,0 +1,260 @@
+    return false;
+}
+
+async function installAmbient(url, filename) {
```
It probably could be made a more generic function instead of copying for all asset types.

Tony-sama replied:
indeed, still need to factorize. Did a crappy copy to test the thing before ending my day.

Tony-sama replied:
Did factorized using one function for any asset. Expect the assets folder to have a folder per type found in the json of ST-content.

Cohee1207 commented on public/scripts/extensions/assets/index.js:
```diff
@@ -0,0 +1,260 @@
+            // Audio
+
+            // Ambient
+            for(const i in availableAssets["audio"]["ambient"]) {
```
Consider refactoring. Get Asset types from JSON and group by types. If needed - we could expand the JSON structure to include needed fields (category, display name, etc.).

Tony-sama replied:
Factorised according to asset type found in the JSON. Reorganised assets folder to have "type" as folder in public/assets, like bgm, ambient. Reduce server.js for audio/assets function to one for getting all installed assets, one to download any file to given path and one for getting character specific assets (used only for bgm for now).

Cohee1207 commented on public/scripts/extensions/audio/index.js:
```diff
@@ -0,0 +1,580 @@
+
+        // 1) Update ambient audio
+        // ---------------------------
+        let newBackground = $("#bg1").css("background-image");
```
There could be also a bg_custom layer (that is set using "Chat Backgrounds" or "Stable Diffusion" plugin). So we need a more generic way of getting a currently selected background (btw, bg_custom is available through chat_metadata object).

Tony-sama replied:
Ty for the tips will look into a cleaner way to get the background :)

Tony-sama replied:
Added a check to handle bg_custom background.

Cohee1207 commented:
> @Tony-sama When will this be merged into staging?

When it's done.

Tony-sama committed:
Merge branch 'SillyTavern:staging' into staging

Tony-sama committed:
Audio extension, handle custome background. Factorised bgm/ambient file listing into one assets listing function on server.js.

Tony-sama committed:
Assets extension, factorised using the received json type to organise the assets in the UI/folders.

Tony-sama committed:
Merge branch 'staging' of https://github.com/Tony-sama/SillyTavern into staging

Tony-sama committed:
Audio group chat cooldown bgm fixed.

Cohee1207 commented on public/scripts/extensions/assets/index.js:
```diff
@@ -0,0 +1,156 @@
+
+async function installAsset(url, assetType, filename) {
+    console.debug(DEBUG_PREFIX,"Downloading ",url);
+    const save_path = "public/assets/"+assetType+"/"+filename;
```
Add assets folder to .gitignore file

Tony-sama replied:
Done

Cohee1207 commented on public/scripts/extensions/assets/index.js:
```diff
@@ -0,0 +1,156 @@
+}
+
+function loadSettings() {
+    updateCurrentAssets().then(function(){
```
I think this should be done on click, with a confirmation. Users won't like ST making internet requests without their permission. And the ability to confirm the link before clicking.

Tony-sama replied:
Added the url of the json as a input field and a connect button next to it. Call popu when clicked, only fetch the json if user confirm. Show toastr error if something went wrong when fetching json (bad url for example).

Cohee1207 commented on public/scripts/extensions/audio/index.js:
```diff
@@ -0,0 +1,581 @@
+//#############################//
+
+const defaultSettings = {
+    enabled: true,
```
Disable this by default please

Tony-sama replied:
changed

Cohee1207 commented on public/scripts/extensions/audio/index.js:
```diff
@@ -0,0 +1,581 @@
+    console.debug(DEBUG_PREFIX, "getting bgm list for", name);
+
+    try {
+        const result = await fetch(`/get_character_assets_list?name=${encodeURIComponent(name)}&assetsFolder=${CHARACTER_BGM_FOLDER}`);
```
GET requests are not immune to CSRF attacks. They should be avoided if possible

Tony-sama replied:
Changed into post (I think, just change into app.post and use header when fetching, is that enough to avoid such attack ?) :)

Cohee1207 replied:
`{ method: 'POST', headers: getRequestHeaders() }` passing to fetch should be enough

Tony-sama replied:
Fine then that's what I did :)

Cohee1207 commented on public/scripts/extensions/audio/index.js:
```diff
@@ -0,0 +1,581 @@
+$(document).ready(function () {
+    function addExtensionControls() {
+        const settingsHtml = `
+        <div id="audio_settings">
```
Take a look at $.get approach instead of HTML strings (Cfg and regex plugins for example). This will save us time for future refactors. Thanks :)

Tony-sama replied:
Changed both assets and audio UI to use the $.get and dedicacated window.html file like in cfg extension. More convenient patern indeed.

Cohee1207 commented on public/scripts/extensions/audio/index.js:
```diff
@@ -0,0 +1,581 @@
+        if(custom_background !== undefined)
+            newBackground = custom_background
+
+        newBackground = newBackground.substring(newBackground.lastIndexOf("/")+1).replace(/\.[^/.]+$/, "").replaceAll("%20","-").replaceAll(" ","-"); // remove path and spaces
```
The background could be a base64 data URI. I think you should exit early in that case (we have a util to detect data URI strings).

Tony-sama replied:
Added a check for base64 URL, such background value should be ignored now.

Cohee1207 commented on public/scripts/extensions/audio/manifest.json:
```diff
@@ -0,0 +1,11 @@
+{
+    "display_name": "Audio",
```
A more engaging name like "Dynamic Audio" could grab more attention. No need to rename the actual folder tho.

Tony-sama replied:
Updated

Cohee1207 commented on server.js:
```diff
@@ -4999,3 +5000,110 @@ app.post('/delete_extension', jsonParser, async (request, response) => {
+
+
+/**
+ * HTTP POST handler function to retrieve name of all files of a given folder path.
```
It's not actually HTTP POST :)

Tony-sama replied:
Indeed, corrected I think.

Cohee1207 commented on server.js:
```diff
@@ -4999,3 +5000,110 @@ app.post('/delete_extension', jsonParser, async (request, response) => {
+});
+
+/**
+ * HTTP POST handler function to retrieve a character background music list.
```
Also not HTTP POST

Tony-sama replied:
Indeed, corrected I think.

Cohee1207 commented on server.js:
```diff
@@ -4999,3 +5000,110 @@ app.post('/delete_extension', jsonParser, async (request, response) => {
+    const res = await fetch(url);
+    const destination = path.resolve(file_path);
+    const fileStream = fs.createWriteStream(destination, { flags: 'wx' });
+    await finished(Readable.fromWeb(res.body).pipe(fileStream));
```
Function is not async, await won't work

Tony-sama replied:
Made it async as well as post I think.

Tony-sama committed:
- Added assets folder to .gitignore
        - disabled audio extension by default
        - turned get request into post
        - avoid background that are data url
        - changed UI name to Dynamic Audio
        - Changed Assets/Audio ui load to use $.get
        - added assets json url as field in ui, with connect button require confirm from popup.

Tony-sama committed:
Merge branch 'SillyTavern:staging' into staging

Tony-sama committed:
update audio manifest

Tony-sama committed:
Changed audio to muted by default to avoid error on autoplay without user interact.

Tony-sama committed:
Added message to Asset UI to tell about need to refresh for asset to be used.

Tony-sama committed:
Added sanitization of request input for assets_download function. Changed assets download UI for button with little animation while downloading.

Tony-sama committed:
Redesigned assets connect button

Tony-sama committed:
Updated assets ui css to make it relative size

Tony-sama committed:
Sanitized character bgm request.

Tony-sama committed:
Merge branch 'SillyTavern:staging' into staging

Tony-sama committed:
Add temp folder into assets, file are download into this temp file and moved only when complete. Allow to handle fail download.

Tony-sama committed:
Merge branch 'staging' of https://github.com/Tony-sama/SillyTavern into staging

Tony-sama committed:
Merge branch 'SillyTavern:staging' into staging

Cohee1207 committed:
Reformat code.

Tony-sama committed:
Merge branch 'SillyTavern:staging' into staging

Tony-sama committed:
Updated audio/assets extension user feedback for muted/disconnected from assets repo.

Cohee1207 committed:
Improve asset downloader

Cohee1207 committed:
Merge branch 'staging' of https://github.com/Tony-sama/SillyTavern into music

Tony-sama committed:
Added a refresh assets button to audio UI to detect audio asset again. Usefull after download an asset or putting a file manually.

Tony-sama committed:
Merge branch 'staging' of https://github.com/Tony-sama/SillyTavern into staging

Tony-sama committed:
Added delete asset api route, factorise asset filename sanitizer to use in both asset download/delete. Added delete function to asset UI download button.

Tony-sama committed:
Merge branch 'SillyTavern:staging' into staging

Tony-sama committed:
Merge branch 'SillyTavern:staging' into staging
